src/Concordium/Client/Runner.hs

```diff
@@ -3450,76 +3452,243 @@ processLegacyCmd action backend =
+        printSuccess
+    DumpStop ->
+      withClient backend $
+        dumpStopV2 >>=
+        printSuccess
+    GetIdentityProviders block ->
+      withClient backend $
+        readBlockHashOrDefault Best block >>=
+        getIdentityProvidersV2 >>=
+        printResponseValueAsJSON
+    GetAnonymityRevokers block ->
+      withClient backend $
+        readBlockHashOrDefault Best block >>=
+        getAnonymityRevokersV2 >>=
+        printResponseValueAsJSON
+    GetCryptographicParameters block ->
+      withClient backend $
+        readBlockHashOrDefault Best block >>=
+        getCryptographicParametersV2 >>=
+        printResponseValueAsJSON
   where
-    printSuccess (Left x)  = liftIO . putStrLn $ x
-    printSuccess (Right x) = liftIO $ if x then putStrLn "OK" else putStrLn "FAIL"
-    printSuccess' = printSuccess . fmap grpcResponseVal
-
--- |Look up block infos all the way to genesis.
-loop :: Either String Value -> ClientMonad IO ()
-loop v =
-  case v of
-    Left err       -> liftIO $ putStrLn err
-    Right (AE.Object m) ->
-      case KM.lookup "blockParent" m of
-        Just (String parent) -> do
-          printJSON v
-          case KM.lookup "blockSlot" m of
-            Just (AE.Number x) | x > 0 ->
-              fmap grpcResponseVal <$> getBlockInfo parent >>= loop
-            _ -> return () -- Genesis block reached.
-        _ -> error "Unexpected return value for block parent."
-    _ -> error "Unexptected return value for getBlockInfo."
-
+    -- |Print the response value under the provided mapping,
+    -- or fail with an error message if the response contained
+    -- an error.
+    printResponseValue :: (MonadIO m, Show b)
+      => (a -> b)
+      -> GRPCResult (Either String a)
+      -> m ()
+    printResponseValue f res =
+      getResponseValueOrFail' f res >>= liftIO . print
+
+    -- |Print the response value as JSON, or fail with an error
+    -- message if the response contained an error.
+    printResponseValueAsJSON :: (MonadIO m, ToJSON a)
+      => GRPCResult (Either String a)
+      -> m ()
+    printResponseValueAsJSON res = do
+      v <- getResponseValueOrFail res
+      printJSONValues . toJSON $ v
+
+    -- |Print result of a query with side-effects.
+    printSuccess (Left x)  = liftIO $ logError [[i|FAIL: #{x}|]]
+    printSuccess (Right _) = liftIO $ logSuccess ["OK"]
+
+    -- |`read` input or fail if the input could not be `read`.
+    readOrFail :: (MonadIO m, Read a) => Text -> m a
+    readOrFail t =
+      case readEither s of
+        Left err -> logFatal [[i|Unable to parse '#{s}': #{err}|]]
+        Right v -> return v
+      where s = Text.unpack t
+
+    -- |Reads a blockhash wrapped in the `Maybe` monad.
+    -- If the provided value is @Nothing@, a default value specified provided
+    -- as the first parameter is returned. If the provided value is string wrapped
+    -- in @Just@, `readOrFail s` is returned. Fails if `s` is not a valid blockhash.
```

Comment:
```suggestion
    -- |If the provided value is @Nothing@, a default value specified provided
    -- as the first parameter is returned. If the provided value is @Just t@ then t
    -- is parsed as a block hash. If this fails 'logFatal' is called to report and terminate the process.
```

The main point I'm trying to get across is that **there is nothing gained by mentioning "monad" here**. There is nothing from the concept of Monad that is relevant for this function.

Reply:
Good point. This feedback is useful for me when writing comments in the future.